zkevm-circuits/src/keccak_circuit/keccak_padding.rs

```diff
@@ -0,0 +1,427 @@
+}
+
+/// KeccakPaddingCircuit
+#[derive(Default)]
+pub struct KeccakPaddingCircuit<F: Field> {
+    inputs: Vec<KeccakPaddingRow<F>>,
+    size: usize,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> KeccakPaddingCircuit<F> {
+    fn r() -> F {
+        F::from(123456)
+    }
+}
+
+impl<F: Field> Circuit<F> for KeccakPaddingCircuit<F> {
+    type Config = KeccakPaddingConfig<F>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        KeccakPaddingConfig::configure(meta)
+    }
+
+    fn synthesize(
+        &self,
+        config: Self::Config,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        config.assign(
+            layouter,
+            self.size,
+            &self.inputs[0],
+            KeccakPaddingCircuit::r(),
+        )?;
+        Ok(())
+    }
+}
+
+impl<F: Field> KeccakPaddingConfig<F> {
+    pub(crate) fn configure(meta: &mut ConstraintSystem<F>) -> Self {
+        let q_enable = meta.selector();
+        let q_end = meta.advice_column();
+        let d_bits = [(); KECCAK_WIDTH].map(|_| meta.advice_column());
+        let d_lens = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let d_rlcs = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let s_flags = [(); KECCAK_RATE_IN_BYTES].map(|_| meta.advice_column());
+        let randomness = meta.advice_column();
+
+        meta.create_gate("boolean checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            //TODO: could be removed if combined with keccak circuit?
+            for data_bit in d_bits {
+                let b = meta.query_advice(data_bit, Rotation::cur());
+                cb.require_boolean("input data bit", b);
+            }
+
+            for s_flag in s_flags {
+                let s = meta.query_advice(s_flag, Rotation::cur());
+                cb.require_boolean("boolean state bit", s);
+            }
+
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+
+                cb.require_boolean("boolean state switch", s_i - s_i_1);
+            }
+
+            cb.gate(meta.query_selector(q_enable))
+        });
+
+        meta.create_gate("padding bit checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
+
+            let mut constraints = vec![];
+            for i in 1..s_flags.len() {
+                let s_i = meta.query_advice(s_flags[i], Rotation::cur());
+                let s_i_1 = meta.query_advice(s_flags[i - 1], Rotation::cur());
+                let d_bit_0 = meta.query_advice(d_bits[8 * i], Rotation::cur());
+                constraints.push(("begin with 1", (s_i - s_i_1) * (d_bit_0 - 1u64.expr())));
```

Comment:
No need for `constraints` I think, can just do these thinks directly on `cb`. For example
```
let s_padding_start = s_i - s_i_1;
cb.condition(s_padding_start, |cb| {
    cb.require_equal(d_bit_0, 1u64.expr());
});
```

Reply:
the require_equal does not return an expression, so I can not do things like (s_i - s_i_1) * cb.require_equal(d_bit_0, 1.expr()). Will see if some APIs could achieve that.

Reply:
fixed!